arch: introduce `arch_stack_walk()` and add implementation for RISCV

ycsin opened the pull request:
While reviewing #72890, I think the stacktrace inplementation in RISCV can probably be factored out so that it can be useful for #72890.

And seems like Linux already has an API for such use case, I guess we can probably just steal it over..

Added an implementation for RISC-V, tested and works on a [modified version](https://github.com/ycsin/zephyr/tree/pr/arch_stack_walk_perf) of #72890
![perf_sp_1](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/bc91ed3f-c6d1-43e2-a926-790b348a4799)

Added a shell command to unwind a stack:
```
uart:~$ kernel threads
Scheduler: 11 since last call
Threads:
*0x80017138 shell_uart
        options: 0x0, priority: 14 timeout: 0
        state: queued, entry: 0x800029ac
        stack size 3072, unused 1316, usage 1756 / 3072 (57 %)

 0x80017ca8 sysworkq
        options: 0x1, priority: -1 timeout: 0
        state: pending, entry: 0x80006842
        stack size 1024, unused 644, usage 380 / 1024 (37 %)

 0x800177e0 idle
        options: 0x1, priority: 15 timeout: 0
        state: , entry: 0x800065ae
        stack size 512, unused 180, usage 332 / 512 (64 %)

 0x80017950 main
        options: 0x1, priority: 0 timeout: 13
        state: suspended, entry: 0x80006326
        stack size 4096, unused 3604, usage 492 / 4096 (12 %)

uart:~$ kernel unwind 0x80017ca8
Unwinding 0x80017ca8 sysworkq
ra: 0x80007114 [z_swap+0x58]
ra: 0x80007ae8 [z_sched_wait+0x10]
ra: 0x8000689a [work_queue_main+0x58]
ra: 0x800006de [z_thread_entry+0x2e]
```

ycsin committed:
arch: introduce `arch_stack_walk()`

An architecture can indicate that it has an implementation for
the `arch_stack_walk()` function by selecting
`ARCH_HAS_STACKWALK`.

Set the default value of `EXCEPTION_STACK_TRACE_MAX_FRAMES` to
`ARCH_STACKWALK_MAX_FRAMES` if the latter is available.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
arch: riscv: stacktrace: implement `arch_stack_walk()`

Created the `arch_stack_walk()` function out from the original
`z_riscv_unwind_stack()`, it's been updated to support
unwinding any thread.

Updated the stack_unwind test case accordingly.

Increased the delay in `test_fatal_on_smp`, to wait
for the the fatal thread to be terminated, as stacktrace can
take a bit more time.

Doubled the kernel/smp testcase timeout from 60 (default) to
120s, as some of the tests can take a little bit more than 60s
to finish.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on include/zephyr/arch/arch_interface.h:
```diff
@@ -1240,6 +1240,35 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
+ * csf		esf	Stack trace starting on esf
+ * ============ ======= ============================================
+ */
+void arch_stack_walk(stack_trace_consume_fn consume_entry, void *cookie,
+		     const void *csf, const void *esf);
```
I'm not sure if this is the right place for this function, not sure if `csf` & `esf` are too low level. They are `const void *` because their type isn't (probably can't be?) available in this file.

ycsin commented on arch/Kconfig:
```diff
@@ -405,6 +405,12 @@ config FRAME_POINTER
 	  increased size and decreased speed.
 
+config ARCH_STACKWALK
+	bool "Architecture-specific stack walk function"
+	depends on ARCH_HAS_STACKWALK
```
probably should default y

andyross commented on include/zephyr/arch/arch_interface.h:
```diff
@@ -46,6 +46,8 @@ typedef struct z_thread_stack_element k_thread_stack_t;
 
 typedef void (*k_thread_entry_t)(void *p1, void *p2, void *p3);
 
+typedef struct __esf z_arch_esf_t;
```
Style: since this is always used in the API by value (and is an arch-specific opaque type anyway) our coding guidelines would demand that it be passed as a struct type.  Since you're cleaning up the API otherwise, might be a good place to fix up.  Maybe call this `struct arch_esf` or `struct arch_exception` or something similar?

I guess we'd need some deprecation strategy for the old typedef; despite being a "z_" internal thing it's always been part of the k_sys_fatal_error_handler() argument list and exposed to apps.

ycsin replied:
I'd go for `struct arch_esf` so that it is more recognizable with the existing `z_arch_esf_t`, or probably just adopt Linux's `struct pt_regs`

andyross commented on include/zephyr/arch/arch_interface.h:
```diff
@@ -1240,6 +1242,35 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
  */
 void arch_spin_relax(void);
 
+#ifdef CONFIG_ARCH_STACKWALK
```
Do we need this as a Kconfig?  Seems like a feature that, if unused, would be dropped from the link.  The only other reason to make it tunable would be to optimize build time for uncalled code, but again this is tiny and IMHO we don't want to be playing games like that at the arch level.

ycsin replied:
Removed `CONFIG_ARCH_STACKWALK` but I retained the `CONFIG_ARCH_HAS_STACKWALK`, which is used to guard the `CONFIG_ARCH_STACKWALK_MAX_FRAMES`. `CONFIG_ARCH_STACKWALK_MAX_FRAMES` (like the `CONFIG_EXCEPTION_STACK_TRACE_MAX_FRAMES`) will be helpful to make sure that we dont stuck in an infinite loop, LMK what you think

andyross commented on include/zephyr/arch/arch_interface.h:
```diff
@@ -1240,6 +1242,35 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
+ *
+ * @return True, if the entry was consumed or skipped. False, if there is no space left to store
+ */
+typedef bool (*stack_trace_consume_fn)(void *cookie, unsigned long addr);
```
So... I get that this is a linux API, so we can presume that it's generally well-designed and understood, yada yada.

That said... what's being "consumed" here?  When you "consume" something it goes away and isn't usable any more.   You can enumerate a dead stack frame in perpetuity, it's not going anywhere and for sure this process doesn't have side effects.   This should be a verb like "visit" or "iterate", no?  Just seems weird.

andyross replied:
Also: do we want to consider adding a parameter for the instruction pointer (e.g. "return_address", "calling_address", depending on convention) here?  Most architectures provide that too in some form (I guess we could allow it to be NULL if unsupported), and for sure anyone using this API will want it.

Not sure how linux provides that: is it another API you call with the stack address as a parameter maybe?

ycsin replied:
`stack_trace_consume_fn` API is actually the same as Linux:

https://github.com/torvalds/linux/blob/cc8ed4d0a8486c7472cd72ec3c19957e509dc68c/include/linux/stacktrace.h#L11-L39

ycsin replied:
I named it `stack_trace_callback_fn ` instead, since it is a callback

andyross reviewed (commented):
Some notes.  Overall this looks clean to me.  Note that there is also a ton of per-arch stack walking code that e.g. does LOG_ERR() or printk() to hand format stack traces that we should look at replacing with a portable one based on this.

ycsin commented on include/zephyr/arch/arch_interface.h:
```diff
@@ -46,6 +46,8 @@ typedef struct z_thread_stack_element k_thread_stack_t;
 
 typedef void (*k_thread_entry_t)(void *p1, void *p2, void *p3);
 
+typedef struct __esf z_arch_esf_t;
```
Unfortunately, not all arch define its esf struct as `struct __esf`, this CI will fail and probably depends on #73593

andyross reviewed (dismissed):
No remaining complaints.  I do think it would be a good idea to at least do a quick audit of the arch layers for ones which could support the walker with an easy port.  And at some point we'll have to solve the way to expose "program counter" as an extension to this API since it's going to be universally demanded anyway (and multiple arch's support it already)

ycsin commented:
> No remaining complaints. I do think it would be a good idea to at least do a quick audit of the arch layers for ones which could support the walker with an easy port. And at some point we'll have to solve the way to expose "program counter" as an extension to this API since it's going to be universally demanded anyway (and multiple arch's support it already)

It is done through the `cookie` pointer in the Linux kernel, right now I pass an `int` to it so that we can continue to print numbered stack traces, but I can probably remove that to not mislead other devs

ycsin commented:
Added implementation for x86 & arm64

ycsin commented:
ping @andyross @dcpleung could you please take a look? Thanks!

dcpleung commented on subsys/debug/Kconfig:
```diff
@@ -382,6 +382,7 @@ config EXCEPTION_STACK_TRACE
 
 config EXCEPTION_STACK_TRACE_MAX_FRAMES
 	int "Configures the depth of stack trace"
+	default ARCH_STACKWALK_MAX_FRAMES
```
I think this also needs `depends on ARCH_HAS_STACKWALK` as `ARCH_STACKWALK_MAX_FRAMES` is not always available.

ycsin replied:
Good catch, fixed, thanks

andyross reviewed (dismissed):
Thanks for the extra arch work

andyross reviewed (dismissed):
Still good with me

ycsin commented:
I was testing this on #72890 and realizes that the stacktrace implementation in RISC-V (and probably the x86 as well) doesn't work probably on something like #72890 because of the stack bounds check

In case this PR can't make it for the LTS, I wonder if we can at least merge the first commit which introduces the API and theoretically allow OOT arch to implement it?

ycsin committed:
arch: riscv: stop printing symbol name at mepc

Now that the unwind starts from mepc already, the symbol
name at the mepc reg is kinda redundant, so just remove it.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

nashif commented:
> In case this PR can't make it for the LTS, I wonder if we can at least merge the first commit which introduces the API and theoretically allow OOT arch to implement it?

No, we do not want to be introducing some interface with no in-tree implementation and arch_ interfaces are not something meant to be implemented out-of-tree. This is an internal interface between kernel and architecture.

ycsin committed:
shell: kernel: add command to unwind a thread

Add a shell command to unwind a thread using its thread id.

uart:~$ kernel threads
Scheduler: 11 since last call
Threads:
*0x80017138 shell_uart
        options: 0x0, priority: 14 timeout: 0
        state: queued, entry: 0x800029ac
        stack size 3072, unused 1316, usage 1756 / 3072 (57 %)

 0x80017ca8 sysworkq
        options: 0x1, priority: -1 timeout: 0
        state: pending, entry: 0x80006842
        stack size 1024, unused 644, usage 380 / 1024 (37 %)

 0x800177e0 idle
        options: 0x1, priority: 15 timeout: 0
        state: , entry: 0x800065ae
        stack size 512, unused 180, usage 332 / 512 (64 %)

 0x80017950 main
        options: 0x1, priority: 0 timeout: 13
        state: suspended, entry: 0x80006326
        stack size 4096, unused 3604, usage 492 / 4096 (12 %)

uart:~$ kernel unwind 0x80017ca8
Unwinding 0x80017ca8 sysworkq
ra: 0x80007114 [z_swap+0x58]
ra: 0x80007ae8 [z_sched_wait+0x10]
ra: 0x8000689a [work_queue_main+0x58]
ra: 0x800006de [z_thread_entry+0x2e]

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Removed the implementation for x86 & arm64 as they probably do not work properly outside of the fatal use case.

ycsin commented:
cc @npitre

ycsin commented:
> arch_ interfaces are not something meant to be implemented out-of-tree

Q: Does this apply to the OOT archs? Do we actually support OOT archs?

npitre commented:
Your commit log says:

>     The fp-based implementation of stack-unwinding when the esf is
>     NULL came from a5d86b8b077466384f2c2ea6771498a1e265bb95.

But...

```
$ git log a5d86b8b077466384f2c2ea6771498a1e265bb95
fatal: bad object a5d86b8b077466384f2c2ea6771498a1e265bb95
```

You may refer to a git hash only if it is merged upstream. Pull requests are
applied with a "rebase" not with a "merge" so any hash from any private tree
will be different once applied.

Then, in the code:

```
       if ((esf == NULL) && (csf == NULL)) {
               return;
       } else if (esf != NULL) {
               fp = esf->s0;
               ra = esf->mepc;
       } else if ((csf == NULL) || csf == &_current->callee_saved) {
               esf = *((struct arch_esf **)(((uintptr_t)_current_cpu->irq_stack) - 16));
               fp = esf->s0;
               ra = (uintptr_t)walk_stackframe;
       } else {
               fp = csf->s0;
               ra = csf->ra;
       }
```

This would require some comments. Way too much magic is going on here.
For instance, what is that `irq_stack - 16` all about? I probably can guess
only because I'm familiar with the IRQ stack switch as I wrote it. But even
so I'm not completely sure. Now imagine anybody else.

Also, commenting how each of these `esf`and `csf` combinations may
come to be would help as well.

Furthermore, if both `esf` and `csf` are `NULL` then the first if is true.
If the second if is false, that means `esf` is null and therefore `csf`
must be non null... meaning the `csf == NULL` condition in the 3rd if will
never be true.

So please don't hesitate to comment such tricky code. Otherwise a year from
now even you won't remember why you wrote it.

ycsin commented:
Thanks for the review

> Furthermore, if both `esf` and `csf` are `NULL` then the first if is true.
> If the second if is false, that means `esf` is null and therefore `csf`
> must be non null... meaning the `csf == NULL` condition in the 3rd if will
> never be true.

Good catch, I've reoredered the if conditionals and added some comments to it, could you please take another look?

ycsin commented:
Increase some of the timeouts in tests as the stacktrace requires a little bit more time than before, and rebased on main

ycsin commented:
I'm not sure if it's just my laptop's performance is bad, the `test_inc_concurrency` alone takes 100+ seconds to finish

```
START - test_inc_concurrency
type 0: cnt 60000, spend 1299 ms
type 1: cnt 60000, spend 53075 ms
type 2: cnt 60000, spend 54327 ms
 PASS - test_inc_concurrency in 108.705 seconds
 ```

Not sure if 120s is enough in upstream

ycsin commented:
ping @andyross @dcpleung, could you please take another look? Thanks
